cpp/src/strings/search/contains_multiple.cu

```diff
@@ -0,0 +1,313 @@
+    // search for byte in first_bytes array
+    auto const sptr     = d_str.data() + str_byte_idx;
+    auto const chr      = static_cast<u_char>(*sptr);
+    auto const byte_ptr = thrust::lower_bound(thrust::seq, d_first_bytes, last_ptr, chr);
+    // if not found, continue to next byte
+    if ((byte_ptr == last_ptr) || (*byte_ptr != chr)) { continue; }
+    // compute index of matched byte
+    auto const offset_idx = static_cast<size_type>(thrust::distance(d_first_bytes, byte_ptr));
+    auto map_idx          = d_offsets[offset_idx];
+    auto const last_idx = (offset_idx + 1) < unique_count ? d_offsets[offset_idx + 1] : num_targets;
+    // check for targets that begin with chr
+    while (map_idx < last_idx) {
+      auto const target_idx = d_indices[map_idx++];
+      auto const bool_idx   = (target_idx * tile_size) + lane_idx;
+      auto const found      = tile_size == 1 ? d_results[target_idx][str_idx] : bools[bool_idx];
+      if (!found) {  // not found before
+        auto const d_target = d_targets.element<string_view>(target_idx);
+        if ((d_str.size_bytes() - str_byte_idx) >= d_target.size_bytes()) {
+          // first char already checked, so just check the [1, end) chars match
+          auto const tp = d_target.data();
+          if (thrust::equal(thrust::seq, tp + 1, tp + d_target.size_bytes(), sptr + 1)) {
+            if constexpr (tile_size == 1) {
+              d_results[target_idx][str_idx] = true;
+            } else {
+              bools[bool_idx] = true;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  if constexpr (tile_size > 1) {
+    __syncwarp();
+    // reduce the bools for each target to store in the result
+    for (auto target_idx = lane_idx; target_idx < num_targets; target_idx += tile_size) {
+      auto const begin = bools + (target_idx * tile_size);
+      d_results[target_idx][str_idx] =
+        thrust::any_of(thrust::seq, begin, begin + tile_size, thrust::identity<bool>{});
+      // cooperative group implementation was almost 3x slower than this parallel reduce
+    }
+  }
+}
+}  // namespace
+
+std::unique_ptr<table> contains_multiple(strings_column_view const& input,
+                                         strings_column_view const& targets,
+                                         rmm::cuda_stream_view stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(
+    not targets.is_empty(), "Must specify at least one target string.", std::invalid_argument);
+  CUDF_EXPECTS(not targets.has_nulls(), "Target strings cannot be null", std::invalid_argument);
+
+  auto const d_strings = column_device_view::create(input.parent(), stream);
+  auto const d_targets = column_device_view::create(targets.parent(), stream);
+
+  // copy the first byte of each target and sort them
+  auto first_bytes = rmm::device_uvector<u_char>(targets.size(), stream);
+  auto indices     = rmm::device_uvector<size_type>(targets.size(), stream);
+  {
```

Comment:
question: what is this scope used for?

Reply:
Just to isolate the cub-sort dependencies and also to free the `tmp_stg` buffer sooner than later.